### PR TITLE
Enhance security and implement cash journal functionality

### PR DESCRIPTION
**Pull request by Fabi911**
- add Backend CRUD for Cash Journal
- add Frontend Income- and Expense Board

### CODE REVIEW

File: backend/src/main/java/de/webdev/backend/cashjournal/service/CashJournalService.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+		return cashJournalRepository.save(booking);
+	}
+
+	public void deleteCashJournal(String id) {
+		cashJournalRepository.deleteById(id);
+	}
+
+	public Booking updateBooking(String id, Booking booking) {
+		return cashJournalRepository.findById(id)
+				.map(existingBooking -> {
```
Comment (esgoet) [inline backend/src/main/java/de/webdev/backend/cashjournal/service/CashJournalService.java]: not sure why you map here if you don't use existingBooking anywhere? could just use `@With` annotation, for example, and then return what you get back with `findById.withDate(booking.date()).withDescription(...` and so on

File: backend/src/main/java/de/webdev/backend/security/AppuserRole.java [hunk 2]
```diff
@@ -2,5 +2,7 @@
 
 public enum AppuserRole {
```
Comment (esgoet) [inline backend/src/main/java/de/webdev/backend/security/AppuserRole.java:3]: not sure how you have it in the rest of your code, but maybe AppUser instead of Appuser?
```suggestion
public enum AppUserRole {
```

File: backend/src/main/java/de/webdev/backend/security/config/SecurityConfig.java [hunk 3]
```diff
@@ -36,10 +37,20 @@ public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Excepti
 						authorizeRequests
 								.requestMatchers(HttpMethod.POST, "api/users/register").permitAll()
 								.requestMatchers(HttpMethod.POST, "api/users/login").permitAll()
+								// Members
 								.requestMatchers(HttpMethod.DELETE, apiMembers).hasRole(AppuserRole.ADMIN.name())
 								.requestMatchers(HttpMethod.GET, apiMembers).hasRole(AppuserRole.ADMIN.name())
 								.requestMatchers(HttpMethod.POST, apiMembers).hasRole(AppuserRole.ADMIN.name())
 								.requestMatchers(HttpMethod.PUT, apiMembers).hasRole(AppuserRole.ADMIN.name())
+								.requestMatchers(HttpMethod.DELETE, apiMembers).hasRole(AppuserRole.GROUP1.name())
```
Comment (esgoet) [inline backend/src/main/java/de/webdev/backend/security/config/SecurityConfig.java:45]: if it's the same for all types of requests, you could just not specify it and only write it once? but obv doesn't hurt, and if you do plan to differentiate between different http request methods, then this makes sense
```suggestion
								.requestMatchers(apiMembers).hasRole(AppuserRole.GROUP1.name())
```

File: backend/src/test/java/de/webdev/backend/cashjournal/controller/CashJournalControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+						    "amount": 101.50,
+						    "category": "Beiträge",
+						    "type": "INCOME"
+						}
+						"""));
+		mockMvc.perform(MockMvcRequestBuilders.get("/api/cash-journal/1"))
+				.andExpect(MockMvcResultMatchers.status().isOk())
+				.andExpect(MockMvcResultMatchers.content().json("""
+						{
+						    "date": "2023-01-22",
```
Comment (esgoet) [inline backend/src/test/java/de/webdev/backend/cashjournal/controller/CashJournalControllerTest.java:75]: would this not also return the id?

Comment (esgoet) [reply backend/src/test/java/de/webdev/backend/cashjournal/controller/CashJournalControllerTest.java:75]: also maybe test for when cash journal by id does not exist?

File: backend/src/test/java/de/webdev/backend/cashjournal/controller/CashJournalControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+						    "description": "Mitgliedsbeiträge",
+						    "amount": 101.50,
+						    "category": "Beiträge",
+						    "type": "INCOME"
+						}
+						"""))
+				.andExpect(MockMvcResultMatchers.status().isOk())
+				.andExpect(MockMvcResultMatchers.content().json("""
+						{
```
Comment (esgoet) [inline backend/src/test/java/de/webdev/backend/cashjournal/controller/CashJournalControllerTest.java:110]: id?

File: backend/src/test/java/de/webdev/backend/cashjournal/controller/CashJournalControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+								"id": "1",
+						    "date": "2023-01-22",
+						    "description": "Mitgliedsbeiträge",
+						    "amount": 101.50,
+						    "category": "Beiträge",
+						    "type": "INCOME"
+						}
+						"""));
+		mockMvc.perform(MockMvcRequestBuilders.delete("/api/cash-journal/1"))
+				.andExpect(MockMvcResultMatchers.status().isOk());
```
Comment (esgoet) [inline backend/src/test/java/de/webdev/backend/cashjournal/controller/CashJournalControllerTest.java:136]: could also test here now that another getAll request is now just returning an empty list

File: backend/src/test/java/de/webdev/backend/cashjournal/controller/CashJournalControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+						    "description": "Mitgliedsbeiträge",
+						    "amount": 101.50,
+						    "category": "Beiträge",
+						    "type": "INCOME"
+						}
+						"""));
+	}
+
+	@Test
+	@WithMockUser(roles = "ADMIN")
```
Comment (esgoet) [inline backend/src/test/java/de/webdev/backend/cashjournal/controller/CashJournalControllerTest.java:85]:
```suggestion
	@WithMockUser(roles = "ADMIN")
	@DirtiesContext
```

File: backend/src/test/java/de/webdev/backend/cashjournal/controller/CashJournalControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+	void getAllCashJournals_shouldReturnEmptyList_whenCallInitially() throws Exception {
+		mockMvc.perform(MockMvcRequestBuilders.get("/api/cash-journal"))
+				.andExpect(MockMvcResultMatchers.status().isOk())
+				.andExpect(MockMvcResultMatchers.content().json("[]"));
+	}
+
+	@Test
+	@WithMockUser(roles = "ADMIN")
```
Comment (esgoet) [inline backend/src/test/java/de/webdev/backend/cashjournal/controller/CashJournalControllerTest.java:32]:
```suggestion
	@WithMockUser(roles = "ADMIN")
	@DirtiesContext
```

File: backend/src/test/java/de/webdev/backend/cashjournal/controller/CashJournalControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+						    "description": "Mitgliedsbeiträge",
+						    "amount": 101.50,
+						    "category": "Beiträge",
+						    "type": "INCOME"
+						}
+						"""));
+	}
+
+	@Test
+	@WithMockUser(roles = "ADMIN")
```
Comment (esgoet) [inline backend/src/test/java/de/webdev/backend/cashjournal/controller/CashJournalControllerTest.java:122]:
```suggestion
	@WithMockUser(roles = "ADMIN")
	@DirtiesContext
```

File: backend/src/test/java/de/webdev/backend/cashjournal/service/CashJournalServiceTest.java
Comment (esgoet) [inline backend/src/test/java/de/webdev/backend/cashjournal/service/CashJournalServiceTest.java:1]: the goal here is to test the CashJournalService methods, so you need to create an instance of CashJournalService and then call those methods in then //When section

File: backend/src/test/java/de/webdev/backend/cashjournal/service/CashJournalServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.Mockito.*;
+
+class CashJournalServiceTest {
+
+	private final CashJournalRepository cashJournalRepository=mock(CashJournalRepository.class);
```
Comment (esgoet) [inline backend/src/test/java/de/webdev/backend/cashjournal/service/CashJournalServiceTest.java]:
```suggestion
	private final CashJournalRepository cashJournalRepository=mock(CashJournalRepository.class);
	private final CashJournalService cashJournalService = new CashJournalService(cashJournalRepository);
```

**Review by esgoet — COMMENTED** [review]
Some @DirtiesContext missing in ControllerTest, and your CashJournalServiceTest needs to be updated, otherwise just optional comments
